launcher/InstanceDirUpdate.cpp

```diff
@@ -0,0 +1,129 @@
+        return QString();
+
+    auto newName = FS::DirNameFromString(instance->name(), QFileInfo(oldRoot).dir().absolutePath());
+    auto newRoot = FS::PathCombine(QFileInfo(oldRoot).dir().absolutePath(), newName);
+    if (oldRoot == newRoot)
+        return QString();
+
+    // Check for conflict
+    if (QDir(newRoot).exists()) {
+        QMessageBox::warning(parent, QObject::tr("Cannot rename instance"),
+                             QObject::tr("New instance root (%1) already exists. <br />Only the metadata will be renamed.").arg(newRoot));
+        return QString();
+    }
+
+    // Ask if we should rename
+    if (renamingMode == "AskEverytime") {
+        QMessageBox messageBox(parent);
+        messageBox.setText(QObject::tr("Would you also like to rename the instance folder?"));
+        messageBox.setInformativeText(QObject::tr("Renaming \'%1\' -> \'%2\'").arg(oldName, newName));
+        messageBox.setStandardButtons(QMessageBox::Yes | QMessageBox::No);
+        messageBox.setDefaultButton(QMessageBox::Yes);
+        messageBox.setIcon(QMessageBox::Question);
+
+        auto checkBox = new QCheckBox(QObject::tr("&Remember my choice"), parent);
+        messageBox.setCheckBox(checkBox);
+
+        auto res = messageBox.exec();
+        if (checkBox->isChecked()) {
+            if (res == QMessageBox::Yes)
+                APPLICATION->settings()->set("InstRenamingMode", "PhysicalDir");
+            else
+                APPLICATION->settings()->set("InstRenamingMode", "MetadataOnly");
+        }
+        if (res == QMessageBox::No)
+            return QString();
+    }
+
+    // Check for linked instances
+    if (!checkLinkedInstances(instance->id(), parent))
+        return QString();
+
+    // Now we can confirm that a renaming is happening
+    if (!instance->syncInstanceDirName(newRoot)) {
+        QMessageBox::warning(parent, QObject::tr("Cannot rename instance"),
+                             QObject::tr("An error occurred when performing the following renaming operation: <br/>"
+                                         " - Old instance root: %1<br/>"
+                                         " - New instance root: %2<br/>"
+                                         "Only the metadata is renamed.")
+                                 .arg(oldRoot, newRoot));
+        return QString();
+    }
+    return newRoot;
+}
+
+bool checkLinkedInstances(const QString& id, QWidget* parent)
+{
+    auto linkedInstances = APPLICATION->instances()->getLinkedInstancesById(id);
+    if (!linkedInstances.empty()) {
+        auto response = CustomMessageBox::selectable(parent, QObject::tr("There are linked instances"),
+                                                     QObject::tr("The following instance(s) might reference files in this instance:\n\n"
+                                                                 "%1\n\n"
+                                                                 "Deleting it could break the other instance(s), \n\n"
```

Comment:
if this is going to reused as a function then "Deleting" should be exchangeable with "Renaming"

Reply:
Sure